### PR TITLE
Implement lab 3 - letters

### LINKED ISSUE #633: Lab 3 - Letter frequency

### PR DESCRIPTION
**Pull request by jimbosoft**
Fixes #633

Review of colleague's PR #534

#### Changes proposed in this PR:
- Implement Lab3 letters
- Test Lab3

### CODE REVIEW

**Review by runnerdave — DISMISSED** [review]
I think is all good, being such a simple exercise, I would recommend you to implement tables for testing though, the Code Owners seem to always want that method instead of the multiple tests you did. Here is explained: [table driven testing](https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go)

File: 03_letters/jimbotech/letter_test.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"bytes"
+	"reflect"
+	"testing"
+)
+
+func TestLetters(t *testing.T) {
+	var sentence = "marry had a little lamb"
```
Comment (camscale) [inline 03_letters/jimbotech/letter_test.go]: nit: "mary" is spelt with one "r".

File: 03_letters/jimbotech/letter_test.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		"\x00:0", "a:93", "ß:100000", "å:-1",
+	}
+	res := sortLetters(testMap)
+
+	eq := reflect.DeepEqual(res, expected)
+	if !eq {
+		t.Errorf("Result %v differs from expected %v", res, expected)
+	}
+}
+func TestEmptySorter(t *testing.T) {
```
Comment (camscale) [inline 03_letters/jimbotech/letter_test.go]: nitpick: blank line between functions.

File: 03_letters/jimbotech/letter_test.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		'a':    93,
+		'ß':    100000,
+		'\x00': 0,
+	}
+	expected := []string{
+		"\x00:0", "a:93", "ß:100000", "å:-1",
+	}
+	res := sortLetters(testMap)
+
+	eq := reflect.DeepEqual(res, expected)
```
Comment (camscale) [inline 03_letters/jimbotech/letter_test.go]: nitpick: idiomatic Go eschews simple variables like this, instead preferring:
```
    if !reflect.DeepEqual(res, expected) {
```

Splitting your `if` over multiple lines by using an intermediate variable actually makes it less readable.

File: 03_letters/jimbotech/letter_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	testMap := map[rune]int{
+		'å':    -1,
+		'a':    93,
+		'ß':    100000,
+		'\x00': 0,
+	}
+	expected := []string{
+		"\x00:0", "a:93", "ß:100000", "å:-1",
+	}
+	res := sortLetters(testMap)
```
Comment (camscale) [inline 03_letters/jimbotech/letter_test.go]: A far more common name for `res` here is `actual`. You'll see that is very common in most tests.

File: 03_letters/jimbotech/letter_test.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		'e': 1,
+		'b': 1,
+		' ': 4,
+	}
+
+	res := letters(sentence)
+
+	eq := reflect.DeepEqual(res, expected)
+	if !eq {
+		t.Errorf("Result maps %v differs from expected %v", res, expected)
```
Comment (camscale) [inline 03_letters/jimbotech/letter_test.go]: Go error strings should not start with capital letters: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

please fix throughout.

File: 03_letters/jimbotech/letter_test.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+	res := sortLetters(testMap)
+
+	if len(res) > 0 {
+		t.Error("Retured map is NOT empty")
+	}
+}
+
+func TestMain(t *testing.T) {
+
```
Comment (camscale) [inline 03_letters/jimbotech/letter_test.go]: nit: I'm not going to go hard on what I see as unnecessary extra blank lines, but please be consistent. All your other functions do not have a blank line after the function signature; this one does.

File: 03_letters/jimbotech/letter_test.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	want := "a:2\nb:1\n"
```
Comment (camscale) [inline 03_letters/jimbotech/letter_test.go]: This is usually called `expected` (along with `actual`). `got` and `want` are also fine, but be consistent. In a previous test you used `expected`, so I *expect* you to also use that here.

File: 03_letters/jimbotech/letter_test.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	var buf bytes.Buffer
+	out = &buf
+	main()
+	result := buf.String()
+
+	if result != want {
+		t.Errorf("expected %v, got %v", want, result)
```
Comment (camscale) [inline 03_letters/jimbotech/letter_test.go]: wow. this is tortured. expected == want, got == result. just use the same name in the code as the string. i.e:
```
t.Errorf("want %v, got %v", want, got)
```
or
```
t.Errorf("expected %v, actual %v", expected, actual)
```

Either way, more consistency please.

**Review by camscale — CHANGES_REQUESTED** [review]
mostly nitpicks which i expect you'll consistently ignore, but again, error strings have a certain requirements in Go for reasons, so at least please fix that.

+1 on sizing the slice capacity in `sortLetters`

File: 03_letters/jimbotech/letter_test.go (hunk 1 above)
Comment (jimbosoft) [reply 03_letters/jimbotech/letter_test.go]: Done

File: 03_letters/jimbotech/letter_test.go (hunk 2 above)
Comment (jimbosoft) [reply 03_letters/jimbotech/letter_test.go]: Done

File: 03_letters/jimbotech/letter_test.go (hunk 3 above)
Comment (jimbosoft) [reply 03_letters/jimbotech/letter_test.go]: Done

File: 03_letters/jimbotech/letter_test.go (hunk 4 above)
Comment (jimbosoft) [reply 03_letters/jimbotech/letter_test.go]: Done

File: 03_letters/jimbotech/letter_test.go (hunk 5 above)
Comment (jimbosoft) [reply 03_letters/jimbotech/letter_test.go]: Done

File: 03_letters/jimbotech/letter_test.go (hunk 6 above)
Comment (jimbosoft) [reply 03_letters/jimbotech/letter_test.go]: Done

File: 03_letters/jimbotech/letter_test.go (hunk 8 above)
Comment (jimbosoft) [reply 03_letters/jimbotech/letter_test.go]: Done

**Review by camscale — APPROVED** [review]
LGTM.

I don't know whether you noticed or not, but I can tell you that reading your tests now that you have consistent "expected" and "actual" naming, it reads quite a lot better to me. I can read it so much faster than before. It may seem like trivial nitpicking, but it does really make a difference. Thanks.

**Pull request merged**